blog: Add "Godot 4.0 sets sail: All aboard for new horizons" (#577)
godotengine/godot-website: collections/_article/godot-4-0-sets-sail.md

Without further ado, feel free to jump straight to the [download page](/download)!

Our friends at [GDQuest](https://www.gdquest.com/) prepared a wonderful video to highlight just how huge the Godot 4.0 release is. They also took the lead on writing this blog post, enabling the contributors from the Production team to focus on technical aspects of this release. The GDQuest team makes amazing [open source demos](https://github.com/gdquest-demos/godot-4.0-new-features) and [YouTube tutorials](https://www.youtube.com/@GDQuest), and they make [professional courses for Godot](https://gdquest.mavenseed.com/courses/)!

<iframe width="560" height="315" src="https://www.youtube.com/embed/chXAjMQrcZk" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

For more details, grab a cup of something strong and enjoy browsing through the massive list of new features.

## Giving back<!-- omit in toc -->

> As a community effort, Godot relies on individual contributors to improve. In recent years, [user and company donations](/donate/) enabled us to hire a number of core contributors to work more hours on the engine, and allow us to finalize the Godot 4.0 release faster. We still need a lot of [extra donation income](/donate/) to be able to sustain the pace we've had for the 4.0 development cycle, and to hire more contributors to focus on key areas which are currently missing a maintainer.

Reviewer: ```suggestion
As a community effort, Godot relies on individual contributors to improve. In recent years, [user and company donations](/donate/) enabled us to also hire a number of core contributors to work more hours on the engine, and allow us to finalize the Godot 4.0 release faster. As a result, our monthly expenses are higher than our monthly income and we have relied on large one-time donations to fund development. Currently, we need a lot more [monthly donations](/donate/) to be able to sustain the pace we've had for the 4.0 development cycle, not to mention the need to hire more contributors to focus on key areas which are currently missing a maintainer.
```

---
title: "Godot 4.0 sets sail: All aboard for new horizons"
excerpt: "It's official. Today marks the beginning of a new era for Godot. After 3+ years of breaking and rebuilding from the ground up, we're thrilled to say: Welcome to the start of Godot 4!" 
categories: ["release"]
author: "2000+ Godot contributors"
image: /storage/blog/covers/godot-4-0-sets-sail.webp

> date: 2023-03-01 17:00:00

Reviewer 2: ```suggestion
date: 2023-03-01 12:00:00
```

	 - [New Rendering Optimization Techniques](#new-rendering-optimization-techniques)
	 - [Enhanced Mid & Post-Processing  ](#enhanced-mid-post-processing-)
 - [2D Improvements](#2d-improvements)
	 - [Powerful New 2D Level-Editing Tools](#powerful-new-2d-level-editing-tools)
	 - [New 2D Rendering Options](#new-2d-rendering-options)
	 - [Improved 2D Lighting & Shadows](#improved-2d-lighting-shadows)
 - [Shaders & VFX](#shaders-vfx)
	 - [New Atmospheric Effects](#new-atmospheric-effects)
	 - [Textures & Material Projection](#textures-material-projection)
	 - [Enhanced Shader-Game World Interaction](#enhanced-shader-game-world-interaction)
	 - [Improved Shader Editor](#improved-shader-editor)
	 - [Extended Shader Language](#extended-shader-language)
	 - [Compute Shaders](#compute-shaders)
 - [Scripting](#scripting)
	 - [GDScript](#gdscript)
	 - [C\#](#c)
	 - [GDExtension - experimental](#gdextension-experimental)
 - [Physics](#physics)
	 - [Game-Specific Physics Engine](#game-specific-physics-engine)
	 - [Multithreading & Performance Optimization](#multithreading-performance-optimization)
	 - [Better Physics API](#better-physics-api)

> 	 - [Higher Simulation Stability](#higher-simulation-stability)
> 	 - [Remarks](#remarks)
>  - [UI & Text](#ui-text)

Reviewer 2: ```suggestion
	 - [Higher Simulation Stability](#higher-simulation-stability)
 - [UI & Text](#ui-text)
```

	 - [Multiple Window Support](#multiple-window-support)
	 - [UI Editor Improvements](#ui-editor-improvements)
	 - [New Text Rendering Systems](#new-text-rendering-systems)
	 - [New Theme & Theme Editor](#new-theme-theme-editor)
 - [Internationalization](#internationalization)
	 - [Extended Language Support](#extended-language-support)
	 - [Easier Translation Workflow](#easier-translation-workflow)
 - [Editor & UX](#editor-ux)
	 - [Easier Importing](#easier-importing)
	 - [New Editor Features & Widgets](#new-editor-features-widgets)
	 - [Inspector Dock Improvements](#inspector-dock-improvements)
	 - [Scene Dock Improvements](#scene-dock-improvements)
	 - [Script Editor Improvements](#script-editor-improvements)
	 - [Easier Version Control](#easier-version-control)
	 - [New Movie Maker Mode](#new-movie-maker-mode)
	 - [New Editor Theme](#new-editor-theme)
 - [Navigation](#navigation)
	 - [Server-Based Navigation System](#server-based-navigation-system)

> 	 - [Extended Complex Navigation Support](#extended-complex-navigation-support)
> 	 - [Important Remarks:](#important-remarks)
>  - [XR](#xr)

Reviewer 2: ```suggestion
	 - [Extended Complex Navigation Support](#extended-complex-navigation-support)
 - [XR](#xr)
```

![GDBot the robot standing face camera in front of two trees](/storage/blog/godot-4-0-sets-sail/02-3d-gi-sdfgi.webp)

GIProbe has been replaced by the **VoxelGI** node, a real-time solution fit for small and medium-scale environments with particularly good results for interiors. Of course, you can still use lightmaps to pre-render lighting and shadows on low-end devices but lightmap baking now uses the GPU for much faster rendering.

![Interior of a tiny but well-lit apartment with only a living room and a kitchen](/storage/blog/godot-4-0-sets-sail/02-3d-gi-voxelgi.webp)

Lastly, we were never quite satisfied with how shadows looked in Godot 3. Godot 4 was an opportunity to go back to the drawing board on shadow rendering to achieve higher quality and provide more granular control.

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-0-sets-sail/02-3d-shadows.mp4" type="video/mp4">
</video>

<a id="new-rendering-optimization-techniques"></a>
### New Rendering Optimization Techniques

> Godot 4 puts at your disposal several new rendering optimization techniques, made possible by Joan Fons ([JFonS](https://github.com/JFonS)) and Juan.  
> The new automatic [occlusion culling](https://github.com/godotengine/godot/pull/48050) can detect models hidden by other geometry and dynamically remove them to increase both CPU and GPU rendering performance.

Reviewer 2: ```suggestion
Godot 4 puts at your disposal several new rendering optimization techniques, made possible by Joan Fons ([JFonS](https://github.com/JFonS)) and Juan.

The new automatic [occlusion culling](https://github.com/godotengine/godot/pull/48050) can detect models hidden by other geometry and dynamically remove them to increase both CPU and GPU rendering performance.
```

---
title: "Godot 4.0 sets sail: All aboard for new horizons"

> excerpt: "Today marks the beginning of a new era for Godot. After 3+ years of breaking and rebuilding from the ground up, we're thrilled to say: Welcome to the start of Godot 4!" 

Reviewer 2: ```suggestion
excerpt: "It's official. Today marks the beginning of a new era for Godot. After 3+ years of breaking and rebuilding from the ground up, we're thrilled to say: Welcome to the start of Godot 4!" 
```

categories: ["release"]
author: "2000+ Godot contributors"
image: /storage/blog/covers/godot-4-0-sets-sail.webp

> date: 2023-02-26 18:00:00

Reviewer 2: Also a reminder to update the date.

```suggestion
date: 2023-03-01 17:00:00
```

---

It's official. Today marks the beginning of a new era for Godot.

> After 3+ years of breaking and rebuilding from the ground up, a complete core overhaul and a full engine rewrite, through 17 alphas, 17 betas and 5 release candidates, we're thrilled to say:

Reviewer: Just a reminder that the number of release candidates will need to change right before releasing

> After 3+ years of breaking and rebuilding from the ground up, a complete core overhaul and a full engine rewrite, through 17 alphas, 17 betas and 5 release candidates, we're thrilled to say:

Reviewer 3: ```suggestion
After 3+ years of breaking and rebuilding from the ground up, a complete core overhaul and a full engine rewrite, through 17 alphas, 17 betas and 6 release candidates, we're thrilled to say:
```

> It's official. Today marks the beginning of a new era for Godot.
> 
> After 3+ years of breaking and rebuilding from the ground up, a complete core overhaul and a full engine rewrite, through 17 alphas, 17 betas and 6 release candidates, we're thrilled to say:
> 
> Welcome to the start of Godot 4! Time to reach new heights together.

Reviewer 2: ```suggestion
After 3+ years of breaking and rebuilding from the ground up, a complete core overhaul and a full engine rewrite, through 17 alphas, 17 betas and 6 release candidates, we're thrilled to say:

**Welcome to the start of Godot 4! Time to reach new heights together.**
```

> We're extremely excited but most of all, we're humbled by the experience. There's no other way to put it: This project is the stuff of case studies as one of the most incredible examples of open collaboration and co-development. So here's to each and every member of our brilliant community of contributors and testers. We're proud to work with you. Godot 4.0 is the culmination of years of your time and effort. Together, we've built a strong new base, one that opens up new horizons and gives Godot the wings it needed to ride the winds of change, to grow with new tech and continue to meet evolving user needs.
> 
> From this point onward, we build on a fresh foundation and focus our upcoming efforts on polishing and performance improvements.
> 
> **As it was with Godot 3.0, Godot 4.0 is only the beginning of Godot 4**. We still expect users to encounter workflow-breaking bugs (especially on less common hardware). Some workflows will still feel somewhat unpolished, and performance won't be optimized yet in this first stable release. But rest assured, we plan to publish releases frequently and regularly (as we have with the Alphas and Betas). So you can expect versions 4.0.1, 4.0.2, etc. to follow very soon.
> 
> We're also working on a publicly available [benchmark suite](https://github.com/godotengine/godot-benchmarks) in which we will be tracking progressive performance changes throughout development. Bit by bit, with your help, we'll close in on our shared vision for Godot 4.
> 
> As a contributor, going forward in the short term, you will find we still won't be looking at rolling in any breaking changes or integrating new features. Those will have to [wait for Godot 4.1 later this year](https://godotengine.org/article/release-management-4-0-and-beyond).
> 
> As for Godot 3 users, needless to say, you'll continue to receive a lot of care as we backport relevant features and bugfixes to the upcoming Godot 3.6. In fact, you will notice a few of the implementations developed for Godot 4 have already made it into Godot 3.5.

Reviewer 2: ```suggestion
We're extremely excited but most of all, we're humbled by the experience. We believe that this project is one of the most incredible examples of open collaboration and co-development. So here's to each and every member of our brilliant community of contributors and testers!

We're proud to work with you. Godot 4.0 is the culmination of years of your time and effort. Together, we've built a strong new base. One that opens up new horizons and gives Godot the wings to ride the winds of change, to grow with new tech and continue to meet evolving user needs. From this point onward, we build on this foundation and focus our upcoming efforts on usability and performance improvements.

**As it was with Godot 3.0, Godot 4.0 is only the beginning of Godot 4**. We still expect users to encounter workflow-breaking bugs (especially on less common hardware). Some workflows will still feel somewhat unpolished, and performance won't be optimized yet in this first stable release. But rest assured, we plan to publish releases frequently and regularly (as we have with the Alphas and Betas). So you can expect versions 4.0.1, 4.0.2, etc. to follow very soon, with new features and bigger improvements [coming later this year in Godot 4.1](https://godotengine.org/article/release-management-4-0-and-beyond).

As for Godot 3 users, needless to say, you'll continue to receive a lot of care as we backport relevant features and bugfixes to the upcoming Godot 3.6. This is going to be our long-term support (LTS) release, that we plan to maintain for the foreseeable future to enable existing Godot 3 project. Throughout the development of Godot 4 we've been backporting a lot of compatible and relevant work, and you will notice a few of new features have already made it into Godot 3.4 and 3.5.
```

Reviewer 3: I'm fine with the edits 👍

![GDBot the robot standing face camera in front of two trees](/storage/blog/godot-4-0-sets-sail/02-3d-gi-sdfgi.webp)

GIProbe has been replaced by the **VoxelGI** node, a real-time solution fit for small and medium-scale environments with particularly good results for interiors. Of course, you can still use lightmaps to pre-render lighting and shadows on low-end devices but lightmap baking now uses the GPU for much faster rendering.

![Interior of a tiny but well-lit apartment with only a living room and a kitchen](/storage/blog/godot-4-0-sets-sail/02-3d-gi-voxelgi.webp)

Lastly, we were never quite satisfied with how shadows looked in Godot 3. Godot 4 was an opportunity to go back to the drawing board on shadow rendering to achieve higher quality and provide more granular control.

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-0-sets-sail/02-3d-shadows.mp4" type="video/mp4">
</video>

<a id="new-rendering-optimization-techniques"></a>
### New Rendering Optimization Techniques

Godot 4 puts at your disposal several new rendering optimization techniques, made possible by Joan Fons ([JFonS](https://github.com/JFonS)) and Juan.  

> In addition to manual [occlusion culling](https://github.com/godotengine/godot/pull/48050) which we backported to Godot 3.5, the new automatic occlusion culling can scan your scene and detect models hidden by other geometry.

Reviewer: The occlusion culling used in 4.0 was not backported to 3.x. Likewise, the occlusion culling in 3.x was not forward-ported to 4.0. 

```suggestion
The new automatic [occlusion culling](https://github.com/godotengine/godot/pull/48050) can detect models hidden by other geometry and dynamically remove them to increase both CPU and GPU rendering performance.
```

Reviewer 2: If this part is edited, please either remove the line break or add another line so it looks like a proper paragraph.

Despite growing in features, the GDScript runtime is much faster and more stable in Godot 4. This was achieved by a complete rewrite of the language backend by our main scripting maintainer George Marques ([vnen](https://github.com/vnen)). If you are interested in further reading, George provided several detailed reports on the new language features ([1](https://godotengine.org/article/gdscript-progress-report-new-gdscript-now-merged), [2](https://godotengine.org/article/gdscript-progress-report-feature-complete-40)) and on the decision-making process for the new language parser and runtime ([1](https://godotengine.org/article/gdscript-progress-report-writing-tokenizer), [2](https://godotengine.org/article/gdscript-progress-report-writing-new-parser), [3](https://godotengine.org/article/gdscript-progress-report-type-checking-back), [4](https://godotengine.org/article/gdscript-progress-report-typed-instructions)).

<a id="c"></a>
### C\#

The much-anticipated port to .NET 6 now allows users to target a newer framework that brings optimizations and new APIs. With .NET 6, projects use C# 10 by default and all features are available.

> Of course, the 4.0 release is also a great opportunity to break compatibility and improve the API. If you’re a C# user we highly recommend checking out this [blog post by Raul Santos, to find out all that’s new in C# for Godot 4.0](https://godotengine.org/article/whats-new-in-csharp-for-godot-4-0/).

Reviewer 2: ```suggestion
Of course, the 4.0 release is also a great opportunity to break compatibility and improve the API. If you’re a C# user we highly recommend checking out this [blog post by Raul Santos](https://godotengine.org/article/whats-new-in-csharp-for-godot-4-0/), to find out all that’s new in C# for Godot 4.0.
```

Links tend to look better if they are only applied to the key part of the sentence, with extra context put outside of it.

Scripting properties is simpler now as well. In previous engine versions, properties related to moving, sliding, and colliding had to be manually passed to move_and_slide(). They can now be set up on the nodes themselves using scenes. This reduces the amount of code needed for desired physical interactions. In addition, physics layers and masks have been made more intuitive.

![](/storage/blog/godot-4-0-sets-sail/06-physics-better-character-body.png)

<a id="higher-simulation-stability"></a>
### Higher Simulation Stability

A new release is not just new features. A significant effort went into fixing previous issues causing jitter and imprecise computations. So you can look forward to higher simulation stability. For more on this effort, contributors Camille Mohr-Daurat ([pouleyKetchoupp](https://github.com/pouleyKetchoupp)), [lawnjelly](https://github.com/lawnjelly), and Fabrice Cipolla ([fabriceci](https://github.com/fabriceci)) have documented their work in [this blog post](https://godotengine.org/article/physics-progress-report-1).

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-0-sets-sail/06-physics-kicking-ball.mp4" type="video/mp4">
</video>

> <a id="remarks"></a>
> ### Remarks

Reviewer 2: ```suggestion
**Important remarks:**
```

Remarks are formatted inconsistently, bold text without a heading is IMO the best option out of available. (This also affects TOC)

For those of you creating apps or games using Arabic Scripts or East Asian logograms, you will find that right-to-left languages work just as you'd expect them to — ligatures, complex graphemes and all. You can read Pāvels' detailed reports on the improvements made: [1](https://godotengine.org/article/complex-text-layouts-progress-report-1), [2](https://godotengine.org/article/complex-text-layouts-progress-report-2), [3](https://godotengine.org/article/complex-text-layouts-progress-report-3).

Besides supporting ligatures, font families and other OpenType features, font resources have two important differences from Godot 3. The first one is proper multilevel fallback logic, which helps to cover a wider range of characters than would be possible with a single font resource. The second is font size variation. The size of the font is no longer tied to the font itself, which means it can be easily changed on the fly. In fact, all Control nodes that have configurable fonts now allow you to vary the size in their theme properties.

<a id="new-theme-theme-editor"></a>
### New Theme & Theme Editor

> Speaking of themes, the new theme editor gives you more control and simplifies your UI design workflow. You can now access your local fonts directly in the editor.  Finally, the default project theme has been modernized to provide a cleaner look and get rid of embedded images, which should slightly reduce the size of exported projects. We have our core contributor Hugo Locurcio ([Calinou](https://github.com/Calinou)) to thank for this. If you've been using Godot 3.5, you will have noted that some of these improvements have already been backported to Godot 3.

Reviewer 2: ```suggestion
Speaking of themes, the new theme editor gives you better tools for creating complex looks and simplifies your UI design workflow. You can now access fonts installed to your local system directly in the editor. Finally, the default project theme has been modernized to provide a cleaner look and get rid of embedded images, which should slightly reduce the size of exported projects. We have our core contributors Hugo Locurcio ([Calinou](https://github.com/Calinou)) and [Yuri Sizov](https://github.com/YuriSizov) to thank for this. If you've been using Godot 3.5, you will have noted that some of these improvements have already been backported to Godot 3.
```

A shameless plug, plus a couple of tweaks.

<a id="extended-complex-navigation-support"></a>
### Extended Complex Navigation Support

The new **NavigationServer** now supports fully dynamic environments and on-the-fly navigation mesh baking. You can stream regions, which makes the system applicable to large open spaces. Physics bodies, whether static or moving, can be marked as obstacles for automatic collision avoidance. It all works faster than before thanks to multithreading support.

In Godot 4.0, [Navigation links](https://github.com/godotengine/godot/pull/63479) allow you to configure jump points, teleports, etc. With this, AI agents can navigate to any point of interest in 2D or 3D scenes, crossing over gaps, walking onto moving platforms, climbing ladders, and more.

![](/storage/blog/godot-4-0-sets-sail/11-navigation-links.webp)

As we have continuously reiterated throughout this article, 4.0 is only the start of the Godot 4 journey. Exciting work is ongoing to further [improve avoidance algorithms to support complex scenarios](https://github.com/godotengine/godot/pull/69988). Upcoming releases will bring additional workflow refinements and more performance optimization.

<a id="important-remarks"></a>

> ### Important Remarks:

Reviewer 2: ```suggestion
**Important remarks:**
```

Remarks are formatted inconsistently, bold text without a heading is IMO the best option out of available. (This also affects TOC)

<a id="more-exporting-options"></a>
### More Exporting Options

Of course, you want your games or apps to be accessible to everyone too. To this end, Godot 4 presents you with new first-class build system support for multiple CPU architectures. If you have the builds compiled, you can now target devices such as Raspberry Pi, Microsoft Volterra, Surface Pro X, Pine Phone, VisionFive, ARM Chromebooks, and Asahi Linux without much manual hassle. This is in addition to the existing support Godot has for x86 Windows & Linux, and various architectures on Android, iOS, and macOS.

We haven't forgotten web-based games. Among many other improvements to the web platform, Fabio has added the necessary tooling to allow [remote profiling of HTML5 exports](https://godotengine.org/article/html5-export-profiling). This means you can run the debugger/profiler on web exports as you would on the desktop platform making it much easier to polish and optimize your web-based games.

One limitation of our philosophy of inclusiveness is that it works best when everyone practices it. That is not the case with console manufacturers, who typically do not open-source their SDKs.

> You can of course create games in Godot 3 for XBox One, PS4, Nintendo Switch, and more with the help of third parties. Several teams are already working on console support for Godot 4, hoping to make it more accessible to everyone.

Reviewer 2: ```suggestion
You can of course create games in Godot 3 for Xbox, PlayStation, Nintendo Switch, and more with the help of third parties. Several teams are already working on console support for Godot 4, hoping to make it more accessible to everyone.
```

Better Xbox capitalization and less specific hardware versions to avoid dating this line too much.
